File: server/scripts/create-recycle-tables.js
/**
 * 创建回收管理数据表
 * 使用 Supabase Management API 执行 SQL
 * 用法: node server/scripts/create-recycle-tables.js
 */
const https = require('https');
const fs = require('fs');
const path = require('path');

function loadEnv() {
  const envFiles = ['.env.local', '.env.development', '.env'];
  for (const file of envFiles) {
    const envPath = path.join(__dirname, '../../', file);
    if (fs.existsSync(envPath)) {
      const content = fs.readFileSync(envPath, 'utf8');
      content.split('\n').forEach(line => {
        const match = line.match(/^([^=]+)=(.*)$/);
        if (match) {
          const key = match[1].trim();
          let value = match[2].trim();
          if (value.startsWith("'") && value.endsWith("'") ||
              value.startsWith('"') && value.endsWith('"')) {
            value = value.slice(1, -1);
          }
          process.env[key] = value;
        }
      });
      console.log(`✅ 已加载环境变量: ${file}`);
      break;
    }
  }
}

function executeSQLViaAPI(sql) {
  return new Promise((resolve, reject) => {
    const PROJECT_REF = process.env.COZE_SUPABASE_URL?.match(/https:\/\/([^.]+)\.supabase\.co/)?.[1];
    const API_KEY = process.env.COZE_SUPABASE_SERVICE_ROLE_KEY || process.env.COZE_SUPABASE_ANON_KEY;

    if (!PROJECT_REF || !API_KEY) {
      reject(new Error('缺少 SUPABASE_URL 或 SERVICE_ROLE_KEY 环境变量'));
      return;
    }

    // 使用 Supabase 的 SQL API
    const options = {
      hostname: `${PROJECT_REF}.supabase.co`,
      port: 443,
      path: '/rest/v1/rpc/execute_sql',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'apikey': API_KEY,
        'Authorization': `Bearer ${API_KEY}`,
        'Prefer': 'return=minimal'
      }
    };

    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve(data);
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
        }
      });
    });

    req.on('error', reject);
    req.write(JSON.stringify({ sql }));
    req.end();
  });
}

async function main() {
  try {
    loadEnv();

    console.log('📋 准备创建回收管理数据表...\n');

    const migrationFile = path.join(__dirname, '../../server/database/migrations/024_create_recycle_management.sql');
    const sql = fs.readFileSync(migrationFile, 'utf8');

    console.log('📝 SQL 内容预览:');
    console.log(sql.substring(0, 300) + '...\n');

    console.log('⚠️  Supabase SDK 不支持直接执行 DDL');
    console.log('尝试使用 API 方法...\n');

    try {
      await executeSQLViaAPI(sql);
      console.log('✅ 数据表创建成功!\n');
    } catch (error) {
      console.log('❌ API 方法失败:', error.message);
      console.log('\n请使用以下方法之一创建表:\n');
      console.log('方法 1: 使用 Supabase Dashboard（推荐）');
      console.log('  1. 打开: https://supabase.com/dashboard/project/xkufzwehvgdmxhutfggh/sql/new');
      console.log('  2. 复制以下文件内容并执行:');
      console.log('     server/database/migrations/024_create_recycle_management.sql\n');
      console.log('方法 2: 使用 psql 命令');
      console.log('  psql $DATABASE_URL -f server/database/migrations/024_create_recycle_management.sql\n');
      process.exit(1);
    }

  } catch (error) {
    console.error('❌ 错误:', error.message);
    process.exit(1);
  }
}

main();

File: server/scripts/create-recycle-tables.py
#!/usr/bin/env python3
"""
创建回收管理数据表
使用 psycopg2 直接连接 PostgreSQL 执行 SQL
用法: python3 server/scripts/create-recycle-tables.py
"""
import os
import sys
import re
from pathlib import Path

# 添加项目根目录到 Python 路径
project_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(project_root))

def load_env():
    """加载环境变量"""
    env_files = ['.env.local', '.env.development', '.env']
    for env_file in env_files:
        env_path = project_root / env_file
        if env_path.exists():
            with open(env_path, 'r', encoding='utf-8') as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith('#') and '=' in line:
                        key, value = line.split('=', 1)
                        key = key.strip()
                        value = value.strip()
                        # 移除引号
                        if value.startswith("'") and value.endswith("'") or \
                           value.startswith('"') and value.endswith('"'):
                            value = value[1:-1]
                        os.environ[key] = value
            print(f"✅ 已加载环境变量: {env_file}")
            break

def parse_supabase_url(url):
    """解析 Supabase URL 获取 PostgreSQL 连接信息"""
    # Supabase URL 格式: https://xkufzwehvgdmxhutfggh.supabase.co
    # PostgreSQL URL 格式: postgresql://postgres:[password]@db.[project-ref].supabase.co:5432/postgres

    match = re.match(r'https://([^.]+)\.supabase\.co', url)
    if not match:
        raise ValueError(f"无效的 Supabase URL: {url}")

    project_ref = match.group(1)

    # 注意: 我们需要数据库密码，这需要从 Supabase Dashboard 获取
    # 由于我们没有密码，我们无法直接连接

    return {
        'host': f'db.{project_ref}.supabase.co',
        'port': 5432,
        'database': 'postgres',
        'user': 'postgres'
    }

def execute_sql_via_python():
    """使用 Python 执行 SQL"""
    try:
        load_env()

        supabase_url = os.environ.get('COZE_SUPABASE_URL') or os.environ.get('SUPABASE_URL')
        if not supabase_url:
            raise ValueError('缺少 COZE_SUPABASE_URL 环境变量')

        # 解析连接信息
        conn_info = parse_supabase_url(supabase_url)

        # 读取 SQL 文件
        migration_file = project_root / 'server/database/migrations/024_create_recycle_management.sql'
        if not migration_file.exists():
            raise FileNotFoundError(f"迁移文件不存在: {migration_file}")

        with open(migration_file, 'r', encoding='utf-8') as f:
            sql = f.read()

        print("📋 准备创建回收管理数据表...")
        print(f"📝 SQL 文件: {migration_file}")
        print(f"📝 SQL 预览: {sql[:200]}...\n")

        print("⚠️  需要数据库密码才能直接连接 PostgreSQL")
        print("请使用以下方法之一创建表:\n")

        print("方法 1: 使用 Supabase Dashboard（推荐）")
        print("  1. 打开浏览器，访问: https://supabase.com/dashboard/project/xkufzwehvgdmxhutfggh/sql/new")
        print("  2. 复制以下文件内容:")
        print(f"     {migration_file}")
        print("  3. 粘贴到 SQL Editor 中")
        print("  4. 点击 Run 执行\n")

        print("方法 2: 使用 psql 命令")
        print(f"  psql $DATABASE_URL -f {migration_file}\n")

        print("方法 3: 如果你有数据库密码，可以使用此脚本:")
        print(f"  export DB_PASSWORD='your-password'")
        print(f"  python3 {__file__}\n")

        # 尝试使用环境变量中的密码
        db_password = os.environ.get('DB_PASSWORD') or os.environ.get('POSTGRES_PASSWORD')
        if db_password:
            import psycopg2
            conn_str = f"postgresql://{conn_info['user']}:{db_password}@{conn_info['host']}:{conn_info['port']}/{conn_info['database']}"
            print(f"🔗 连接数据库: {conn_str[:50]}...")

            conn = psycopg2.connect(conn_str)
            conn.autocommit = True
            cursor = conn.cursor()

            print("⏳ 执行 SQL...")
            cursor.execute(sql)

            print("✅ 数据表创建成功!")

            cursor.close()
            conn.close()
            return True
        else:
            print("💡 提示: 设置 DB_PASSWORD 环境变量后，此脚本可以自动创建表")
            return False

    except Exception as e:
        print(f"❌ 错误: {e}")
        import traceback
        traceback.print_exc()
        return False

if __name__ == '__main__':
    success = execute_sql_via_python()
    sys.exit(0 if success else 1)
